File: cms-app-crawler-ts/src/utils/SelectorRegistry.ts
import fs from 'fs';
import path from 'path';

let _fallbackMap: Map<string, string[]> = new Map();
let _loaded = false;

const REGISTRY_PATH = path.join(__dirname, '../../resources/pom_registry.json');

function load(registryPath: string = REGISTRY_PATH): void {
  if (_loaded) return;
  if (!fs.existsSync(registryPath)) {
    console.warn(`SelectorRegistry: registry not found at ${registryPath} — no fallbacks available`);
    _loaded = true;
    return;
  }
  const entries: Array<{ selectorKey?: string; selectorFallbacks?: string[] }> =
    JSON.parse(fs.readFileSync(registryPath, 'utf8'));
  _fallbackMap.clear();
  for (const entry of entries) {
    const key = entry.selectorKey;
    if (!key) continue;
    _fallbackMap.set(key, entry.selectorFallbacks ?? []);
  }
  _loaded = true;
  console.info(`SelectorRegistry loaded: ${_fallbackMap.size} entries from ${registryPath}`);
}

function getFallbacks(selectorKey: string): string[] {
  if (!_loaded) load();
  return _fallbackMap.get(selectorKey) ?? [];
}

function reset(): void {
  _fallbackMap.clear();
  _loaded = false;
}

export const SelectorRegistry = { load, getFallbacks, reset };

File: cms-app-crawler-py/src/utils/selector_registry.py
import json
import logging
import os
from pathlib import Path

logger = logging.getLogger(__name__)

_fallback_map: dict[str, list[str]] = {}
_loaded = False

_REGISTRY_PATH = Path(__file__).parent.parent.parent / "resources" / "pom_registry.json"


def _load(registry_path: Path = _REGISTRY_PATH) -> None:
    global _fallback_map, _loaded
    if _loaded:
        return
    if not registry_path.exists():
        logger.warning("SelectorRegistry: registry not found at %s — no fallbacks available", registry_path)
        _loaded = True
        return
    with open(registry_path, encoding="utf-8") as f:
        entries: list[dict] = json.load(f)
    _fallback_map.clear()
    for entry in entries:
        key = entry.get("selectorKey")
        if not key:
            continue
        _fallback_map[key] = entry.get("selectorFallbacks", [])
    _loaded = True
    logger.info("SelectorRegistry loaded: %d entries from %s", len(_fallback_map), registry_path)


class SelectorRegistry:
    @staticmethod
    def load(registry_path: str | None = None) -> None:
        path = Path(registry_path) if registry_path else _REGISTRY_PATH
        _load(path)

    @staticmethod
    def get_fallbacks(selector_key: str) -> list[str]:
        if not _loaded:
            _load()
        return _fallback_map.get(selector_key, [])

    @staticmethod
    def reset() -> None:
        global _fallback_map, _loaded
        _fallback_map = {}
        _loaded = False
